algosdk/logic.py

```diff
@@ -0,0 +1,122 @@
+import json
+import os
+
+from . import error
+
+spec = None
+opcodes = None
+
+MAX_COST = 20000
+MAX_LENGTH = 1000
+
+def check_program(program, args=None):
+    """
+    Performs program checking for max length and cost
+
+    Args:
+        program (bytes)
+        args (list[bytes])
+
+    Returns:
+        True on success
+        Raises InvalidProgram on error
+    """
+
+    global spec, opcodes
+
+    if not program:
+        raise error.InvalidProgram("empty program")
+    if not args:
+        args = []
+
+    if spec is None:
+        script_path = os.path.realpath(__file__)
+        script_dir = os.path.dirname(script_path)
+        langspec_file = os.path.join(script_dir, "data", "langspec.json")
+        with open(langspec_file, "rt") as fin:
+            spec = json.load(fin)
+
+    version, vlen = parse_uvariant(program)
+    if vlen <= 0 or version > spec["EvalMaxVersion"]:
+        raise error.InvalidProgram("unsupported version")
+
+    cost = 0
+    length = len(program)
+    for arg in args:
+        length += len(arg)
+
+    if length > MAX_LENGTH:
+        raise error.InvalidProgram("program too long")
+
+    if opcodes is None:
+        opcodes = dict()
+        for op in spec['Ops']:
+            opcodes[op['Opcode']] = op
```

Comment:
opcodes = {opcode: op for op, opcode in spec['Ops'].items()}

Reply:
this does not work, Ops is an array of dicts, but anyway this is a matter of opinion.